server/src/main/java/org/opensearch/tasks/TaskCancellationMonitoringService.java

```diff
@@ -0,0 +1,202 @@
+
+    public TaskCancellationMonitoringService(
+        ThreadPool threadPool,
+        TaskManager taskManager,
+        TaskResourceTrackingService taskResourceTrackingService,
+        TaskCancellationMonitoringSettings taskCancellationMonitoringSettings
+    ) {
+        this.threadPool = threadPool;
+        this.taskManager = taskManager;
+        this.taskCancellationMonitoringSettings = taskCancellationMonitoringSettings;
+        this.cancelledTaskTracker = new ConcurrentHashMap<>();
+        cancellationStatsHolder = TASKS_TO_TRACK.stream()
+            .collect(Collectors.toConcurrentMap(task -> task, task -> new TaskCancellationStatsHolder()));
+        taskManager.addTaskCancellationListeners(this);
+        taskResourceTrackingService.addTaskCompletionListener(this);
+    }
+
+    void doRun() {
+        if (!taskCancellationMonitoringSettings.isEnabled() || !shouldRun.get()) {
+            return;
+        }
+        Map<Class<? extends CancellableTask>, List<CancellableTask>> taskCancellationListByType = getCurrentRunningTasksPostCancellation();
+        taskCancellationListByType.forEach((key, value) -> {
+            long uniqueTasksRunningCount = value.stream().filter(task -> {
+                if (this.cancelledTaskTracker.containsKey(task.getId()) && !this.cancelledTaskTracker.get(task.getId())) {
+                    // Mark it as seen by the stats logic.
+                    this.cancelledTaskTracker.put(task.getId(), true);
+                    return true;
+                } else {
+                    return false;
+                }
+            }).count();
+            cancellationStatsHolder.get(key).totalLongRunningCancelledTaskCount.inc(uniqueTasksRunningCount);
+        });
+    }
+
+    @Override
+    protected void doStart() {
+        scheduledFuture = threadPool.scheduleWithFixedDelay(() -> {
+            try {
+                doRun();
+            } catch (Exception e) {
+                logger.debug("Exception occurred in Task monitoring service", e);
+            }
+        }, taskCancellationMonitoringSettings.getInterval(), ThreadPool.Names.GENERIC);
+    }
+
+    @Override
+    protected void doStop() {
+        if (scheduledFuture != null) {
+            scheduledFuture.cancel();
+        }
+    }
+
+    @Override
+    protected void doClose() throws IOException {
+
+    }
+
+    // For testing
+    protected Map<Long, Boolean> getCancelledTaskTracker() {
+        return this.cancelledTaskTracker;
+    }
+
+    // For testing
+    protected boolean shouldRun() {
+        return shouldRun.get();
+    }
+
+    /**
+     * Invoked when a task is completed. This helps us to disable monitoring service when there are no cancelled tasks
+     * running to avoid wasteful work.
+     * @param task task which got completed.
+     */
+    @Override
+    public void onTaskCompleted(Task task) {
+        if (!TASKS_TO_TRACK.contains(task.getClass())) {
+            return;
```

Comment:
onTaskCompleted gets invoked for every task (whether it completes or gets cancelled) to avoid leaking any map entries

Reply:
Yes it gets called for every task during TaskManager.unregister.